File: dart/14_03_23.dart
class Solution {
  int res = 0;
  int sumNumbers(TreeNode? root) {
    if (root == null) return 0;
    dfs(root, root.val.toString());
    return res;
  }

  dfs(TreeNode node, String currentVal) {
    if (node.left == null && node.right == null) {
      res += int.parse(currentVal);
      return;
    }
    if (node.left != null)
      dfs(node.left!, currentVal + node.left!.val.toString());
    if (node.right != null)
      dfs(node.right!, currentVal + node.right!.val.toString());
  }
}

void main(List<String> args) {
  TreeNode root =
      TreeNode(4, TreeNode(9, TreeNode(5), TreeNode(1)), TreeNode(0));
  print(Solution().sumNumbers(root));
}

class TreeNode {
  int val;
  TreeNode? left;
  TreeNode? right;
  TreeNode([this.val = 0, this.left, this.right]);
}

File: python/14_03_23.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def sumNumbers(self, root):
        if not root: return 0
        self.res = 0
        self.dfs(root, str(root.val))
        return self.res

    def dfs(self, node, cur_str):
        if not node.left and not node.right:
            self.res += int(cur_str)
            return 
        
        if node.left: self.dfs(node.left, cur_str + str(node.left.val))
        if node.right: self.dfs(node.right, cur_str + str(node.right.val))



root = TreeNode(4, TreeNode(9, TreeNode(5), TreeNode(1)), TreeNode(0));
print(Solution().sumNumbers(root));

# Copied from github with modification